### PR TITLE
Template Part block: Fall back to current theme if no theme attribute is given.

### PR DESCRIPTION
**Pull request by ockham**
Part of the task in WordPress Core seeking performance improvements:
- https://github.com/WordPress/wordpress-develop/pull/5455.
- https://core.trac.wordpress.org/ticket/59583

An alternative approach to #53423.

Result of a [discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1696959836535969?thread_ts=1696880738.627909&cid=C02RQBWTW) with @felixarntz, aiming to fix the performance regression re-introduced by https://core.trac.wordpress.org/changeset/56818.

## What?
If a Template Part block is encountered that doesn't have a `theme` attribute, fall back to the current them (`get_stylesheet()`) instead.

## Why?
This should allow removing injection of the `theme` attribute during rendering of the Template Part block.

## How?
By falling back to the current theme (`get_stylesheet()`) if `$attributes['theme']` isn't set.

## Testing Instructions
Verify that Template Parts blocks that are part of patterns still work, both on the frontend and in the editor. Specifically, verify that there are no `Template part has been deleted or is unavailable` errors.

_Will be followed up by a Core PR to remove `theme` attribute injection on the frontend._

cc/ @gziolo

### CODE REVIEW

File: packages/block-library/src/template-part/index.php
Comment (felixarntz) [inline packages/block-library/src/template-part/index.php]:
```suggestion
	if ( isset( $attributes['theme'] ) ) {
```
😁

File: packages/block-library/src/template-part/index.php
Comment (felixarntz) [inline packages/block-library/src/template-part/index.php]: We probably want to keep this condition to not change the behavior:
```suggestion
	if ( isset( $attributes['slug'] ) && get_stylesheet() === $theme ) {
```

**Review by felixarntz — COMMENTED** [review]
I know this is a WIP, just sharing very early feedback here :)

File: packages/block-library/src/template-part/index.php
Comment (gziolo) [reply packages/block-library/src/template-part/index.php]: Fixed with https://github.com/WordPress/gutenberg/pull/55217/commits/06af1954236e2a8e918dde411b69fe7c7281e47a.

**Comment by gziolo** [comment]
After looking at how https://github.com/WordPress/gutenberg/pull/55217 is implemented, I think we should also remove the following line that get executed only in the Gutenberg plugin:

https://github.com/WordPress/gutenberg/blob/8a9cb70272edcfdfd34f3557848949eb0fd29921/packages/block-library/src/pattern/index.php#L47

The same logic should be handled now on the front end with the fallback logic for the `theme` attribute in the Template Part block.

**Review by mikachan — COMMENTED** [review]
Thank you for working on this! I can confirm this fixes the original issue in https://github.com/WordPress/gutenberg/issues/53194.

I tested using the same themes (Overlaid and Lativ) as when I tested https://github.com/WordPress/gutenberg/pull/53423, and this PR works well 👍

**Comment by pbking** [comment]
Tested in all scenarios I can dream up and works as I would expect.  My only suggestion is that, as @gziolo suggested, the call to `_inject_theme_attribute_in_block_template_content()` be removed in this change.  It's unnecessary with this change.  

Once that change has been made this is ready to ship.

**Review by felixarntz — APPROVED** [review]
Pending the additional feedback in https://github.com/WordPress/gutenberg/pull/55217#issuecomment-1757019650, the change so far looks good to me.

**Comment by felixarntz** [comment]
Given @ockham is out today, I took the liberty of pushing the additionally requested change in https://github.com/WordPress/gutenberg/pull/55217/commits/47608ff7a68aa60d3ed878314a0d57e8d93e8a27. @gziolo @pbking @mikachan Could you please take another look?

**Review by mikachan — APPROVED** [review]
LGTM and works well in my testing, thanks @felixarntz!

**Review by felixarntz — CHANGES_REQUESTED** [review]
I am currently benchmarking and profiling this change further, and it seems the approach here is not right in terms of performance at least. Marking as request for changes for now, let's hold off merging at this point.

**Review by felixarntz — APPROVED** [review]
I have concluded my performance investigation here and left a summary in https://core.trac.wordpress.org/ticket/59583#comment:20.

TL;DR This fix makes sense and is good to merge. However, it doesn't fully address the performance issues (partly because the initial "performance win" was mostly due to the bug introduced.

I have identified another very promising performance opportunity related to this https://core.trac.wordpress.org/ticket/59600 and https://github.com/WordPress/wordpress-develop/pull/5463

**Comment by felixarntz** [comment]
@ockham @gziolo @mikachan I think this is good to merge. I'll leave that to you since I'm not sure on the workflow in relation to backporting to the 6.4 release, but LGTM from my end. Once backported to core, we can update https://github.com/WordPress/wordpress-develop/pull/5455 accordingly.

**Comment by ockham** [comment]
Thanks all! I'll go ahead and merge this then; also going to mark for inclusion in WP RC.

@mikachan We need this in Core to unblock https://github.com/WordPress/wordpress-develop/pull/5455. In principle, it's probably fine to land on Monday, but it might be nice to do it even earlier to give us some more time to discover any potential issues with this. Are y'all planning a package sync this week still by any chance? 😊

File: packages/block-library/src/template-part/index.php
Comment (ockham) [inline packages/block-library/src/template-part/index.php]: 🙈

**Comment by mikachan** [comment]
I just cherry-picked this PR to the 6.4-rc1 branch to get it included in the next release: fd15da3b83

**Pull request merged**
